### ISSUE TITLE
Proxying websockets should be supported

### ISSUE THREAD

**Issue by briansmith** [runconduit/conduit#195]
* We should document the current state of websockets support; does it work? Is there some explicit configuration (e.g. disabling ports) required?
* We should have tests that verify it is possible to inject Conduit into a pod that is a websockets server without any explicit configuration, without disrupting the websockets connections.
* We should have tests that verify it is possible for inject Conduit into a pod that is a websockets client--i.e. one that consumes another websockets server (or, potentially, its own)--without any explicit configuration, without disrupting the websockets connections.

IIRC, Websockets only works for HTTP/1.1; the draft for Websockets over HTTP/2 was never completed to become a standard and I don't think anything implements websockets over HTTP/2. The Websockets spec itself is defined to require HTTP/1.1 in the request & status lines so HTTP/1.0 isn't supported.

/cc @seanmonstar

**Comment by briansmith** [comment]
Maybe I'm missing something but this seems pretty straightforward implementation-wise. Basically when we see that a connection gets "upgraded" to a WebSockets connection then we should remove that connection from the HTTP load balancing pool and change the connection's protocol type from HTTP to "TCP" or a new "Websockets" connection type. The same applies to `CONNECT` handling, probably.

I think we can defer Websockets-specific telemetry to a separate issue (that's not filed yet).

**Comment by seanmonstar** [comment]
This is blocked on an API from hyper. There doesn't yet exist a way to take a socket back if trying to upgrade to new a protocol. Part of the reason for that is wanting to match an upgrade and `CONNECT` API, since they seem very similar, but at the same time, `Upgrade` is done away with in HTTP/2, while `CONNECT` does exist in HTTP/2 but without actually giving up the socket (just reserving the stream).

Related hyper issue: https://github.com/hyperium/hyper/issues/1395

(I wouldn't expect this to be ready by Friday, so I'd nominate removing it back out of the 0.3 milestone.)

**Comment by briansmith** [comment]
I thought of something else that complicates this more than I initially thought: Either we can't enqueue any requests behind a "switch to websockets request" on a connection, or we need to be able to transfer all such enqueued requests to another backend request queue, and/or we need to implement a global (not per-backend) queue (at least for websockets and eventually CONNECT). It seems pretty painful. So I would be OK to remove it from 0.3.

That said, I do think that it would be very nice if we could eliminate all the caveats w.r.t. injecting the Conduit proxy that we had to add to the documentation in 0.2. It looks like this is going to be one of the last remaining caveats. IMO "no caveats" is valuable enough to prioritize this higher than we'd normally specifically prioritize websockets support.

**Comment by wmorgan** [comment]
Let's tentatively move this to 0.4. Agree that removing caveats is important, but sounds like we need some more scoping before we can tackle this.

**Comment by oivindoh** [comment]
*Monitoring;
